<a href="https://colab.research.google.com/github/Irina-64/AI_University/blob/main/%D0%91%D0%B0%D0%B7%D0%B0_%D0%B7%D0%BD%D0%B0%D0%BD%D0%B8%D0%B9_%7C_%D0%9F%D0%B5%D1%80%D0%B5%D0%BC%D0%B5%D0%BD%D0%BD%D1%8B%D0%B5_Python_%7C_%D0%A3%D0%98%D0%98.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Переменные

Переменные содержат данные. Благодаря этому к ним можно обращаться, заново использовать, вызывать и назначать их значения, и так далее.

Понимать переменные важно для работы с любой логикой в программировании. Это то, что знает любой программист вне зависимости от языка программирования, и поэтому это так важно для начинающих.

Простейшее определение переменной — это именованный контейнер для данных, к которым нужно обращаться в программе. Есть две основные причины для этого:

1. Зачастую данные — это больше, чем 2 символа.
2. Обычно к ним нужно обращаться по несколько раз.

##Где используются

* Сложно представить программирование без переменных.Это можно было бы сравнить с обращением к людям, когда при этом не используются их имена и фамилии. Представьте, насколько это было бы неудобно.
* В программировании переменные используются для хранения и обращения к данным разных размеров, типов и форм.

##Правила использования

* При присваивании значения переменная всегда находится слева, а данные — справа.
* Имя переменной может начинаться с символа или нижнего подчеркивания (_).
* Для вывода значения переменной с помощью функции `print()` нужно передавать имя переменной без кавычек.




##Частые ошибки


* Начинать имя переменной с цифры.
* Использовать специальные символы (кроме нижнего подчеркивания) в имени переменной.
* Использовать в переменной пробелы.

##Примеры использования переменных

Предположим, что есть большое число: 149597970. Оно обозначает расстояние между Солнцем и Землей в километрах.

Теперь к этому значению нужно обратиться для выполнения вычислений. Вместо того, чтобы каждый раз вводить его или писать в программе, можно просто использовать переменную. Вот так:

In [ ]:
sun_to_earth = 149597970                  # Назначение переменной 
sun_to_earth = sun_to_earth + 1           # Обращение к значению переменной, вычисление и назначение обновленного значения
print(sun_to_earth)                       # Вывод значения переменной

149597971


Эти данные могут быть гораздо более объёмными. В данных, например, может содержаться 100 числовых значений, а обращаться к  таким данным, вероятно, потребуется в десятках разных строк кода. Таким образом, польза от сохранения длинного значения в именованной ячейке (переменной) очевидна.

Разберем код из примера выше:

1. Сначала создается переменная **sun_to_earth** и ей присваивается значение **149597970**.
2. Переменной **sun_to_earth** присваивается новое значение: **149597971**, полученное в результате вычисления - добавления **1** к ранее сохраненному значению переменной **sun_to_earth**, то есть **149597970 + 1 = 149597971**.
3. Выполняется функция `print()`, которая выводит текущее значение переменной **sun_to_earth**, то есть **149597971**.

Еще примеры:

In [ ]:
dogs_name = 'Рэкс'
dogs_kind = 'Лабрадор'
print('Имя собаки: ', dogs_name, 'Порода: ', dogs_kind, sep='\n')

Имя собаки: 
Рэкс
Порода: 
Лабрадор


Это отличный пример применения не только переменной, но и функции `print()`, ведь в последней используются сразу текст и значение переменной. Здесь  имена переменных выбраны удачно, то есть - они отражают суть сохранённых в них значений.



В одной строке можно присвоить сразу несколько переменных. Пример:

In [ ]:
i, j, k = 'Hello', 55, 21.0765
print(i, j, k)

Hello 55 21.0765


##Советы по использованию переменных:

* Название переменной всегда расположено слева при присваивании. А значение, которое нужно присвоить, располагается справа.
* В **Python** (и любом другом языке программирования) очень важно правильно выбирать название переменной, ведь к ней придется часто обращаться:
* Кое-какие имена являются зарезервированными в **Python**, поэтому для переменных их использовать нельзя: **print, true, false, count, sum** и другие. Со временем вы все их узнаете.
* Названия переменных должны начинаться с символа алфавита или нижнего подчеркивания.
* Если начать переменную с символа или нижнего подчеркивания, то после в нее можно включать и цифры.
* Названия переменных чувствительны к регистру. Если был использован символ в верхнем регистре, то именно так и нужно обращаться к переменой в будущем.

##Практические соглашения именования переменных:

* Можно добавлять небольшой суффикс, чтобы не путать переменную с ключевым словом **Python**: **var_number, var_true, v_start, v_close, the_class, the_one, the_name**. Креативность разрешена, главное, чтобы система оставалась понятной.
* Лучше не использовать переменную в один символ, только если это цикл или любая другая короткая операция. Такие имена как **a, x, y, z** начинают путать уже спустя пару минут после того, как их впервые использовали.
* Слова рекомендуется разделять нижним подчеркиванием: **cars_сounted, doors_сlosed, dogs_saved, ships_fixed**, и так далее. Это повышает читаемость переменных. Сравните: **timeinmoscow** и **time_in_moscow**.
* Рекомендуется (часто требуется) использовать латиницу и английские слова. Так ваш код будут понимать разработчики из других стран.

##Типичные ошибки при работе с переменными

Порядок следования инструкций в коде с переменными играет огромное значение. Переменная должна быть определена ***до того***, как будет использована. Ниже пример ошибки, которую очень часто допускают новички:

In [ ]:
print(greeting)
greeting = 'Father!'

NameError: ignored

Запуск программы, выше завершается с ошибкой

```
 NameError: name 'greeting' is not defined
```
Это ошибка обращения, она означает, что в коде используется имя (иначе говорят - идентификатор), который не определено. Причём в самой ошибке об этом указано прямо:
```
'greeting' is not defined.
```

Кроме неправильного порядка определения, в **Python** встречаются простые опечатки, причём как при использовании переменной, так и при её объявлении.

Количество подобных ошибок уменьшается за счет использования правильно настроенного редактора. Такой редактор подсвечивает имена, которые используются без объявления, и предупреждает о возможных проблемах.